### DISCUSSION TITLE
MeshReflectorMaterial -> Portal Camera

### DISCUSSION THREAD

**Discussion by oveddan** [pmndrs/drei#840] [Q&A]
I'd like to achieve something similar to the MeshReflectorMaterial, but have it behave as a portal camera, somewhat like these demos:

https://medium.com/machine-words/creating-portals-in-three-js-944c4f7104ec
https://threejs.org/examples/?q=portal#webgl_portal

Basically it would be the same as MeshReflectorMaterial except youd configure the virtual camera position and look at.

It would be nice if [the reflector code](https://github.com/pmndrs/drei/blob/0cd48695de2e4d6cda491fde4814dfb432376007/src/core/MeshReflectorMaterial.tsx) was slightly better commented so it would be easier for us less familiar with 3d projections and normals to figure out what's going on and how to modify the code to our needs.

What would we need to change in this code to have the virtual camera have its position and look at configured?

**Comment by mwmwmw** [comment]
There are a few ways to go about it. 

The MeshReflectorMaterial creates a virtual camera to do the render. But, you could pass a reference to a camera into a "renderCamera" property. 

You can add that to the props on this line https://github.com/pmndrs/drei/blob/0cd48695de2e4d6cda491fde4814dfb432376007/src/core/MeshReflectorMaterial.tsx#L25

Then replace the virtualCamera in the mesh reflector with your camera. 

Then you could create something like `<portalMaterial renderCamera={yourCameraRef} />` then you retain complete control of that camera.

The projection code will be tricky to work out if you don't have any 3D projection experience. The Three example you shared has a `renderPortal` function that shows exactly what to do. I would recommend not modifying the `<meshReflectorMaterial/>` but instead create a new material based on the Three example.

**Reply by mwmwmw** [reply]
~This isn't quite ready yet. The perspective of the camera in the portal is off, despite being almost a direct copy of the Three example. I think my camera setup is causing issues.~

https://codesandbox.io/s/portals-jyx6oc?file=/src/Portal.js

But, this should at least give you some idea of what you need to do.

**Reply by mwmwmw** [reply]
I updated the example and it's now perspective correct.

**Reply by oveddan** [reply]
thank you for putting this together!!! it says sandbox not found when clicking the link?

**Reply by mwmwmw** [reply]
Lol, I routinely forget to make it public. Unlocked!

**Reply by oveddan** [reply]
this is an amazing example, thank you!!!!

i'm curious if you could also show....how to make it so that the portal mesh is double sided (i.e. material doublesided). I tried this but when looking through the other side, the image is vertically flipped.  Whats the easiest way to flip the camera for the back side?

**Reply by mwmwmw** [reply]
I don't know how generalized this is (I haven't tested it at all)

https://codesandbox.io/s/portals-double-sided-iy5941

But basically, to render the portal through the backside, you just have to skip the Z inversion. 

To do that I calculate the angle between the portal and the camera, and if it's less than a certain number of degrees (the value seems like a magic number), ignore the Z coordinate. 

At the very extreme angles you'll see a flicker. I think this is pretty common. I think mesh reflector also has this issue. (I remember seeing something about it in the code). I think it fades to black at the angles. 

Sure looks pretty!
